# Getting Started With Pandas

We will begin by introducing the `Series`, `DataFrame`, and `Index` classes, which are the basic building blocks of the pandas library, and showing how to work with them. By the end of this section, you will be able to create DataFrames and perform operations on them to inspect and filter the data.

More information [here](https://pandas.pydata.org/docs/user_guide/index.html)

## Anatomy of a DataFrame

A **DataFrame** is composed of one or more **Series**. The names of the **Series** form the column names, and the row labels form the **Index**.

In [1]:
import pandas as pd

meteorites = pd.read_csv('data/Meteorite_Landings.csv', nrows=5)
meteorites

name   id nametype     recclass  mass (g)  fall  \
0    Aachen    1    Valid           L5        21  Fell   
1    Aarhus    2    Valid           H6       720  Fell   
2      Abee    6    Valid          EH4    107000  Fell   
3  Acapulco   10    Valid  Acapulcoite      1914  Fell   
4   Achiras  370    Valid           L6       780  Fell   

                     year    reclat    reclong           GeoLocation  
0  01/01/1880 12:00:00 AM  50.77500    6.08333     (50.775, 6.08333)  
1  01/01/1951 12:00:00 AM  56.18333   10.23333  (56.18333, 10.23333)  
2  01/01/1952 12:00:00 AM  54.21667 -113.00000    (54.21667, -113.0)  
3  01/01/1976 12:00:00 AM  16.88333  -99.90000     (16.88333, -99.9)  
4  01/01/1902 12:00:00 AM -33.16667  -64.95000   (-33.16667, -64.95)

*Source: [NASA's Open Data Portal](https://data.nasa.gov/Space-Science/Meteorite-Landings/gh4g-9sfh)*

#### Series:

In [2]:
meteorites.name

0      Aachen
1      Aarhus
2        Abee
3    Acapulco
4     Achiras
Name: name, dtype: object

#### Columns:

In [3]:
meteorites.columns

Index(['name', 'id', 'nametype', 'recclass', 'mass (g)', 'fall', 'year',
       'reclat', 'reclong', 'GeoLocation'],
      dtype='object')

#### Index:

In [4]:
meteorites.index

RangeIndex(start=0, stop=5, step=1)

## Creating DataFrames

We can create DataFrames from a variety of sources such as other Python objects, flat files, webscraping, and API requests. Here, we will see just a couple of examples, but be sure to check out [this page](https://pandas.pydata.org/pandas-docs/stable/user_guide/io.html) in the documentation for a complete list.

### Using a flat file

In [5]:
import pandas as pd

meteorites = pd.read_csv('data/Meteorite_Landings.csv')

*Tip: There are many parameters to this function to handle some initial processing while reading in the file &ndash; be sure check out the [documentation](https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html).*

### Using data from an API

Collect the data from [NASA's Open Data Portal](https://data.nasa.gov/Space-Science/Meteorite-Landings/gh4g-9sfh) using the Socrata Open Data API (SODA) with the `requests` library:

In [6]:
import requests

response = requests.get(
    'https://data.nasa.gov/resource/gh4g-9sfh.json',
    params={'$limit': 50_000}
)

if response.ok:
    payload = response.json()
else:
    print(f'Request was not successful and returned code: {response.status_code}.')
    payload = None

In [7]:
payload

[{'name': 'Aachen',
  'id': '1',
  'nametype': 'Valid',
  'recclass': 'L5',
  'mass': '21',
  'fall': 'Fell',
  'year': '1880-01-01T00:00:00.000',
  'reclat': '50.775000',
  'reclong': '6.083330',
  'geolocation': {'latitude': '50.775', 'longitude': '6.08333'}},
 {'name': 'Aarhus',
  'id': '2',
  'nametype': 'Valid',
  'recclass': 'H6',
  'mass': '720',
  'fall': 'Fell',
  'year': '1951-01-01T00:00:00.000',
  'reclat': '56.183330',
  'reclong': '10.233330',
  'geolocation': {'latitude': '56.18333', 'longitude': '10.23333'}},
 {'name': 'Abee',
  'id': '6',
  'nametype': 'Valid',
  'recclass': 'EH4',
  'mass': '107000',
  'fall': 'Fell',
  'year': '1952-01-01T00:00:00.000',
  'reclat': '54.216670',
  'reclong': '-113.000000',
  'geolocation': {'latitude': '54.21667', 'longitude': '-113.0'}},
 {'name': 'Acapulco',
  'id': '10',
  'nametype': 'Valid',
  'recclass': 'Acapulcoite',
  'mass': '1914',
  'fall': 'Fell',
  'year': '1976-01-01T00:00:00.000',
  'reclat': '16.883330',
  'reclong': 

Create the DataFrame with the resulting payload:

In [8]:
import pandas as pd

df = pd.DataFrame(payload)
df.head(3)

name id nametype recclass    mass  fall                     year  \
0  Aachen  1    Valid       L5      21  Fell  1880-01-01T00:00:00.000   
1  Aarhus  2    Valid       H6     720  Fell  1951-01-01T00:00:00.000   
2    Abee  6    Valid      EH4  107000  Fell  1952-01-01T00:00:00.000   

      reclat      reclong                                        geolocation  \
0  50.775000     6.083330     {'latitude': '50.775', 'longitude': '6.08333'}   
1  56.183330    10.233330  {'latitude': '56.18333', 'longitude': '10.23333'}   
2  54.216670  -113.000000    {'latitude': '54.21667', 'longitude': '-113.0'}   

  :@computed_region_cbhk_fwbd :@computed_region_nnqa_25f4  
0                         NaN                         NaN  
1                         NaN                         NaN  
2                         NaN                         NaN

*Tip: `df.to_csv('data.csv')` writes this data to a new file called `data.csv`.*

## Inspecting the data
Now that we have some data, we need to perform an initial inspection of it. This gives us information on what the data looks like, how many rows/columns there are, and how much data we have. 

Let's inspect the `meteorites` data.

#### How many rows and columns are there?

In [9]:
meteorites.shape

(45716, 10)

#### What are the column names?

In [10]:
meteorites.columns

Index(['name', 'id', 'nametype', 'recclass', 'mass (g)', 'fall', 'year',
       'reclat', 'reclong', 'GeoLocation'],
      dtype='object')

#### What type of data does each column currently hold?

In [11]:
meteorites.dtypes

name            object
id               int64
nametype        object
recclass        object
mass (g)       float64
fall            object
year            object
reclat         float64
reclong        float64
GeoLocation     object
dtype: object

#### What does the data look like?

In [13]:
meteorites.head(20)

name   id nametype      recclass  mass (g)  fall  \
0                Aachen    1    Valid            L5      21.0  Fell   
1                Aarhus    2    Valid            H6     720.0  Fell   
2                  Abee    6    Valid           EH4  107000.0  Fell   
3              Acapulco   10    Valid   Acapulcoite    1914.0  Fell   
4               Achiras  370    Valid            L6     780.0  Fell   
5              Adhi Kot  379    Valid           EH4    4239.0  Fell   
6   Adzhi-Bogdo (stone)  390    Valid         LL3-6     910.0  Fell   
7                  Agen  392    Valid            H5   30000.0  Fell   
8                Aguada  398    Valid            L6    1620.0  Fell   
9         Aguila Blanca  417    Valid             L    1440.0  Fell   
10     Aioun el Atrouss  423    Valid  Diogenite-pm    1000.0  Fell   
11                  Aïr  424    Valid            L6   24000.0  Fell   
12      Aire-sur-la-Lys  425    Valid       Unknown       NaN  Fell   
13                Akaba  426    Valid            L6     779.0  Fell   
14             Akbarpur  427    Valid            H4    1800.0  Fell   
15              Akwanga  432    Valid             H    3000.0  Fell   
16              Akyumak  433    Valid     Iron, IVA   50000.0  Fell   
17              Al Rais  446    Valid        CR2-an     160.0  Fell   
18            Al Zarnkh  447    Valid           LL5     700.0  Fell   
19                Alais  448    Valid           CI1    6000.0  Fell   

                      year    reclat    reclong           GeoLocation  
0   01/01/1880 12:00:00 AM  50.77500    6.08333     (50.775, 6.08333)  
1   01/01/1951 12:00:00 AM  56.18333   10.23333  (56.18333, 10.23333)  
2   01/01/1952 12:00:00 AM  54.21667 -113.00000    (54.21667, -113.0)  
3   01/01/1976 12:00:00 AM  16.88333  -99.90000     (16.88333, -99.9)  
4   01/01/1902 12:00:00 AM -33.16667  -64.95000   (-33.16667, -64.95)  
5   01/01/1919 12:00:00 AM  32.10000   71.80000          (32.1, 71.8)  
6   01/01/1949 12:00:00 AM  44.83333   95.16667  (44.83333, 95.16667)  
7   01/01/1814 12:00:00 AM  44.21667    0.61667   (44.21667, 0.61667)  
8   01/01/1930 12:00:00 AM -31.60000  -65.23333    (-31.6, -65.23333)  
9   01/01/1920 12:00:00 AM -30.86667  -64.55000   (-30.86667, -64.55)  
10  01/01/1974 12:00:00 AM  16.39806   -9.57028  (16.39806, -9.57028)  
11  01/01/1925 12:00:00 AM  19.08333    8.38333   (19.08333, 8.38333)  
12  01/01/1769 12:00:00 AM  50.66667    2.33333   (50.66667, 2.33333)  
13  01/01/1949 12:00:00 AM  29.51667   35.05000     (29.51667, 35.05)  
14  01/01/1838 12:00:00 AM  29.71667   77.95000     (29.71667, 77.95)  
15  01/01/1959 12:00:00 AM   8.91667    8.43333    (8.91667, 8.43333)  
16  01/01/1981 12:00:00 AM  39.91667   42.81667  (39.91667, 42.81667)  
17  01/01/1957 12:00:00 AM  24.41667   39.51667  (24.41667, 39.51667)  
18  01/01/2001 12:00:00 AM  13.66033   28.96000     (13.66033, 28.96)  
19  01/01/1806 12:00:00 AM  44.11667    4.08333   (44.11667, 4.08333)

Sometimes there may be extraneous data at the end of the file, so checking the bottom few rows is also important:

In [15]:
meteorites.tail(20)

name     id nametype              recclass   mass (g)  \
45696           Zakłodzie  30390    Valid        Enst achon-ung     8680.0   
45697          Zapaliname  30392    Valid          Iron, IAB-MG    85000.0   
45698       Zapata County  30393    Valid                  Iron        NaN   
45699  Zapotitlán Salinas  30394    Valid                    L4       27.7   
45700            Zaragoza  48916    Valid          Iron, IVA-an   162000.0   
45701              Zegdou  30398    Valid                    H3     6700.0   
45702             Zelfana  31353    Valid                    L5     1058.0   
45703               Zenda  30400    Valid     Iron, IAB complex     3700.0   
45704               Zerga  30402    Valid                   LL6       76.0   
45705            Zerhamra  30403    Valid        Iron, IIIAB-an   630000.0   
45706             Zerkaly  31354    Valid                    H5    16000.0   
45707            Zhaoping  54609    Valid     Iron, IAB complex  2000000.0   
45708            Zhigansk  30405    Valid           Iron, IIIAB   900000.0   
45709          Zhongxiang  30406    Valid                  Iron   100000.0   
45710          Zillah 001  31355    Valid                    L6     1475.0   
45711          Zillah 002  31356    Valid               Eucrite      172.0   
45712              Zinder  30409    Valid  Pallasite, ungrouped       46.0   
45713                Zlin  30410    Valid                    H4        3.3   
45714           Zubkovsky  31357    Valid                    L6     2167.0   
45715          Zulu Queen  30414    Valid                  L3.7      200.0   

        fall                    year    reclat    reclong  \
45696  Found  01/01/1998 12:00:00 AM  50.76278   22.86611   
45697  Found  01/01/1998 12:00:00 AM  25.00944 -100.75000   
45698  Found  01/01/1930 12:00:00 AM  27.00000  -99.00000   
45699  Found  01/01/1984 12:00:00 AM  18.33333  -97.50000   
45700  Found                     NaN  41.65000   -0.86667   
45701  Found  01/01/1998 12:00:00 AM  29.75000   -4.50000   
45702  Found  01/01/2002 12:00:00 AM  32.15833    4.63333   
45703  Found  01/01/1955 12:00:00 AM  42.51333  -88.48944   
45704  Found  01/01/1973 12:00:00 AM  20.25000  -12.68333   
45705  Found  01/01/1967 12:00:00 AM  29.85861   -2.64500   
45706  Found  01/01/1956 12:00:00 AM  52.13333   81.96667   
45707  Found  01/01/1983 12:00:00 AM  24.23333  111.18333   
45708  Found  01/01/1966 12:00:00 AM  68.00000  128.30000   
45709  Found  01/01/1981 12:00:00 AM  31.20000  112.50000   
45710  Found  01/01/1990 12:00:00 AM  29.03700   17.01850   
45711  Found  01/01/1990 12:00:00 AM  29.03700   17.01850   
45712  Found  01/01/1999 12:00:00 AM  13.78333    8.96667   
45713  Found  01/01/1939 12:00:00 AM  49.25000   17.66667   
45714  Found  01/01/2003 12:00:00 AM  49.78917   41.50460   
45715  Found  01/01/1976 12:00:00 AM  33.98333 -115.68333   

                  GeoLocation  
45696    (50.76278, 22.86611)  
45697     (25.00944, -100.75)  
45698           (27.0, -99.0)  
45699       (18.33333, -97.5)  
45700       (41.65, -0.86667)  
45701           (29.75, -4.5)  
45702     (32.15833, 4.63333)  
45703   (42.51333, -88.48944)  
45704      (20.25, -12.68333)  
45705      (29.85861, -2.645)  
45706    (52.13333, 81.96667)  
45707   (24.23333, 111.18333)  
45708           (68.0, 128.3)  
45709           (31.2, 112.5)  
45710       (29.037, 17.0185)  
45711       (29.037, 17.0185)  
45712     (13.78333, 8.96667)  
45713       (49.25, 17.66667)  
45714     (49.78917, 41.5046)  
45715  (33.98333, -115.68333)

#### Get some information about the DataFrame

In [16]:
meteorites.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45716 entries, 0 to 45715
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   name         45716 non-null  object 
 1   id           45716 non-null  int64  
 2   nametype     45716 non-null  object 
 3   recclass     45716 non-null  object 
 4   mass (g)     45585 non-null  float64
 5   fall         45716 non-null  object 
 6   year         45425 non-null  object 
 7   reclat       38401 non-null  float64
 8   reclong      38401 non-null  float64
 9   GeoLocation  38401 non-null  object 
dtypes: float64(3), int64(1), object(6)
memory usage: 3.5+ MB


###### [Exercise 1.1](../exercises/pandas_intro_exercise.ipynb#Exercise-1.1)

Create a DataFrame by reading in the `2019_Yellow_Taxi_Trip_Data.csv` file. Examine the first 5 rows.

###### [Exercise 1.2](../exercises/pandas_intro_exercise.ipynb#Exercise-1.2)

Find the dimensions (number of rows and number of columns) in the data.

## Extracting subsets

A crucial part of working with DataFrames is extracting subsets of the data: finding rows that meet a certain set of criteria, isolating columns/rows of interest, etc. After narrowing down our data, we are closer to discovering insights. This section will be the backbone of many analysis tasks.

#### Selecting columns

We can select columns as attributes if their names would be valid Python variables:

In [17]:
meteorites.name

0            Aachen
1            Aarhus
2              Abee
3          Acapulco
4           Achiras
            ...    
45711    Zillah 002
45712        Zinder
45713          Zlin
45714     Zubkovsky
45715    Zulu Queen
Name: name, Length: 45716, dtype: object

If they aren't, we have to select them as keys. However, we can select multiple columns at once this way:

In [18]:
meteorites[['name', 'mass (g)']]

name  mass (g)
0          Aachen      21.0
1          Aarhus     720.0
2            Abee  107000.0
3        Acapulco    1914.0
4         Achiras     780.0
...           ...       ...
45711  Zillah 002     172.0
45712      Zinder      46.0
45713        Zlin       3.3
45714   Zubkovsky    2167.0
45715  Zulu Queen     200.0

[45716 rows x 2 columns]

#### Selecting rows

In [19]:
meteorites[100:104]

name     id nametype       recclass  mass (g)  fall  \
100         Benton   5026    Valid            LL6    2840.0  Fell   
101         Berduc  48975    Valid             L6     270.0  Fell   
102         Béréba   5028    Valid  Eucrite-mmict   18000.0  Fell   
103  Berlanguillas   5029    Valid             L6    1440.0  Fell   

                       year    reclat   reclong          GeoLocation  
100  01/01/1949 12:00:00 AM  45.95000 -67.55000      (45.95, -67.55)  
101  01/01/2008 12:00:00 AM -31.91000 -58.32833  (-31.91, -58.32833)  
102  01/01/1924 12:00:00 AM  11.65000  -3.65000       (11.65, -3.65)  
103  01/01/1811 12:00:00 AM  41.68333  -3.80000     (41.68333, -3.8)

#### Indexing

We use `iloc[]` to select rows and columns by their position:

In [20]:
meteorites.iloc[100:104, [0, 3, 4, 6]]

name       recclass  mass (g)                    year
100         Benton            LL6    2840.0  01/01/1949 12:00:00 AM
101         Berduc             L6     270.0  01/01/2008 12:00:00 AM
102         Béréba  Eucrite-mmict   18000.0  01/01/1924 12:00:00 AM
103  Berlanguillas             L6    1440.0  01/01/1811 12:00:00 AM

We use `loc[]` to select by name:

In [21]:
meteorites.loc[100:104, 'mass (g)':'year']

mass (g)  fall                    year
100    2840.0  Fell  01/01/1949 12:00:00 AM
101     270.0  Fell  01/01/2008 12:00:00 AM
102   18000.0  Fell  01/01/1924 12:00:00 AM
103    1440.0  Fell  01/01/1811 12:00:00 AM
104     960.0  Fell  01/01/2004 12:00:00 AM

#### Filtering with Boolean masks

A **Boolean mask** is a array-like structure of Boolean values &ndash; it's a way to specify which rows/columns we want to select (`True`) and which we don't (`False`).

Here's an example of a Boolean mask for meteorites weighing more than 50 grams that were found on Earth (i.e., they were not observed falling):

In [22]:
(meteorites['mass (g)'] > 50) & (meteorites.fall == 'Found')

0        False
1        False
2        False
3        False
4        False
         ...  
45711     True
45712    False
45713    False
45714     True
45715     True
Length: 45716, dtype: bool

**Important**: Take note of the syntax here. We surround each condition with parentheses, and we use bitwise operators (`&`, `|`, `~`) instead of logical operators (`and`, `or`, `not`).

We can use a Boolean mask to select the subset of meteorites weighing more than 1 million grams (1,000 kilograms or roughly 2,205 pounds) that were observed falling:

In [23]:
meteorites[(meteorites['mass (g)'] > 1e6) & (meteorites.fall == 'Fell')]

name     id nametype    recclass    mass (g)  fall  \
29         Allende   2278    Valid         CV3   2000000.0  Fell   
419          Jilin  12171    Valid          H5   4000000.0  Fell   
506  Kunya-Urgench  12379    Valid          H5   1100000.0  Fell   
707  Norton County  17922    Valid     Aubrite   1100000.0  Fell   
920   Sikhote-Alin  23593    Valid  Iron, IIAB  23000000.0  Fell   

                       year    reclat    reclong             GeoLocation  
29   01/01/1969 12:00:00 AM  26.96667 -105.31667  (26.96667, -105.31667)  
419  01/01/1976 12:00:00 AM  44.05000  126.16667      (44.05, 126.16667)  
506  01/01/1998 12:00:00 AM  42.25000   59.20000           (42.25, 59.2)  
707  01/01/1948 12:00:00 AM  39.68333  -99.86667   (39.68333, -99.86667)  
920  01/01/1947 12:00:00 AM  46.16000  134.65333      (46.16, 134.65333)

*Tip: Boolean masks can be used with `loc[]` and `iloc[]`.*

An alternative to this is the `query()` method:

In [ ]:
meteorites.query("`mass (g)` > 1e6 and fall == 'Fell'")

*Tip: Here, we can use both logical operators and bitwise operators.*

## Calculating summary statistics

In the next section of this workshop, we will discuss data cleaning for a more meaningful analysis of our datasets; however, we can already extract some interesting insights from the `meteorites` data by calculating summary statistics.

In [24]:
meteorites.fall.value_counts()

fall
Found    44609
Fell      1107
Name: count, dtype: int64

In [26]:
meteorites['fall'].value_counts()

fall
Found    44609
Fell      1107
Name: count, dtype: int64

#### How many of the meteorites were found versus observed falling?

*Tip: Pass in `normalize=True` to see this result as percentages. Check the [documentation](https://pandas.pydata.org/docs/reference/api/pandas.Series.value_counts.html) for additional functionality.*

#### What was the mass of the average meterorite?

In [27]:
meteorites['mass (g)'].median()

32.6

We can take this a step further and look at quantiles:

In [28]:
meteorites['mass (g)'].quantile([0.01, 0.05, 0.95, 0.99])

0.01        0.44
0.05        1.10
0.95     4000.00
0.99    50600.00
Name: mass (g), dtype: float64

#### What was the mass of the heaviest meteorite?

In [29]:
meteorites['mass (g)'].max()

60000000.0

Let's extract the information on this meteorite:

In [30]:
meteorites.loc[meteorites['mass (g)'].idxmax()]

name                             Hoba
id                              11890
nametype                        Valid
recclass                    Iron, IVB
mass (g)                   60000000.0
fall                            Found
year           01/01/1920 12:00:00 AM
reclat                      -19.58333
reclong                      17.91667
GeoLocation     (-19.58333, 17.91667)
Name: 16392, dtype: object

#### How many different types of meteorite classes are represented in this dataset?

In [31]:
meteorites.recclass.nunique()

466

Some examples:

In [32]:
meteorites.recclass.unique()[:14]

array(['L5', 'H6', 'EH4', 'Acapulcoite', 'L6', 'LL3-6', 'H5', 'L',
       'Diogenite-pm', 'Unknown', 'H4', 'H', 'Iron, IVA', 'CR2-an'],
      dtype=object)

*Note: All fields preceded with "rec" are the values recommended by The Meteoritical Society. Check out [this Wikipedia article](https://en.wikipedia.org/wiki/Meteorite_classification) for some information on meteorite classes.*

#### Get some summary statistics on the data itself
We can get common summary statistics for all columns at once. By default, this will only be numeric columns, but here, we will summarize everything together:

In [33]:
meteorites.describe(include='all')

name            id nametype recclass      mass (g)   fall  \
count    45716  45716.000000    45716    45716  4.558500e+04  45716   
unique   45716           NaN        2      466           NaN      2   
top     Aachen           NaN    Valid       L6           NaN  Found   
freq         1           NaN    45641     8285           NaN  44609   
mean       NaN  26889.735104      NaN      NaN  1.327808e+04    NaN   
std        NaN  16860.683030      NaN      NaN  5.749889e+05    NaN   
min        NaN      1.000000      NaN      NaN  0.000000e+00    NaN   
25%        NaN  12688.750000      NaN      NaN  7.200000e+00    NaN   
50%        NaN  24261.500000      NaN      NaN  3.260000e+01    NaN   
75%        NaN  40656.750000      NaN      NaN  2.026000e+02    NaN   
max        NaN  57458.000000      NaN      NaN  6.000000e+07    NaN   

                          year        reclat       reclong GeoLocation  
count                    45425  38401.000000  38401.000000       38401  
unique                     266           NaN           NaN       17100  
top     01/01/2003 12:00:00 AM           NaN           NaN  (0.0, 0.0)  
freq                      3323           NaN           NaN        6214  
mean                       NaN    -39.122580     61.074319         NaN  
std                        NaN     46.378511     80.647298         NaN  
min                        NaN    -87.366670   -165.433330         NaN  
25%                        NaN    -76.714240      0.000000         NaN  
50%                        NaN    -71.500000     35.666670         NaN  
75%                        NaN      0.000000    157.166670         NaN  
max                        NaN     81.166670    354.473330         NaN

**Important**: `NaN` values signify missing data. For instance, the `fall` column contains strings, so there is no value for `mean`; likewise, `mass (g)` is numeric, so we don't have entries for the categorical summary statistics (`unique`, `top`, `freq`).

#### Check out the documentation for more descriptive statistics:

- [Series](https://pandas.pydata.org/docs/reference/series.html#computations-descriptive-stats)
- [DataFrame](https://pandas.pydata.org/docs/reference/frame.html#computations-descriptive-stats)

##### [Exercise 1.3](../exercises/pandas_intro_exercise.ipynb#Exercise-1.3)

Using the data in the `2019_Yellow_Taxi_Trip_Data.csv` file, calculate summary statistics for the `fare_amount`, `tip_amount`, `tolls_amount`, and `total_amount` columns.

##### [Exercise 1.4](../exercises/pandas_intro_exercise.ipynb#Exercise-1.4)
Find the dimensions (number of rows and number of columns) in the data.

return to [overview](../00_overview.ipynb)